### PR TITLE
feat(opsem): add an intrinsic sea_assert_if

### PR DESCRIPTION
**Pull request by priyasiddharth**

### CODE REVIEW

File: include/seahorn/Analysis/SeaBuiltinsInfo.hh [hunk 1]
```diff
@@ -17,6 +17,7 @@ enum class SeaBuiltinsOp {
   ASSERT,             /* verifier.assert */
   ASSERT_NOT,         /* verifier.assert.not */
   IS_DEREFERENCEABLE, /* sea.is_dereferenceable */
+  ASSERT_IF,          /* sea.assert.if */
```
Comment (agurfinkel) [inline include/seahorn/Analysis/SeaBuiltinsInfo.hh:20]: would be useful to also add `sea.assert.not.if`. I use the `not` version of the functions to avoid creating extra negation instruction in llvm

File: lib/Analysis/SeaBuiltinsInfo.cc [hunk 2]
```diff
@@ -134,6 +137,26 @@ Function *SeaBuiltinsInfo::mkIsDereferenceable(Module &M) {
...
+Function *SeaBuiltinsInfo::mkAssertIf(llvm::Module &M) {
+  auto &C = M.getContext();
+  auto FC = M.getOrInsertFunction(SEA_ASSERT_IF, Type::getInt1Ty(C), Type::getInt1Ty(C), Type::getInt1Ty(C));
+  auto *FN = dyn_cast<Function>(FC.getCallee());
+  if (FN) {
+    FN->setOnlyAccessesInaccessibleMemory();
+    FN->setDoesNotThrow();
+    FN->setDoesNotFreeMemory();
+    FN->setDoesNotRecurse();
+    // TODO: why does nocapture work in sea_is_def and not here?
```
Comment (agurfinkel) [inline lib/Analysis/SeaBuiltinsInfo.cc:149]: this function takes no pointers so it has no ability to capture anything. Hence it cannot have `nocapture` attribute.

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc [hunk 3]
```diff
@@ -173,6 +174,18 @@ bool PromoteVerifierCalls::runOnFunction(Function &F) {
...
 
+      I.replaceAllUsesWith(ci);
+      toKill.push_back(&I);
+    } else if (fn && (fn->getName().equals("sea_assert_if"))) {  // sea_assert_if is the user facing name
+      IRBuilder<> Builder(F.getContext());
+      Builder.SetInsertPoint(&I);
+      // arg0: antecedent
+      // arg1: consequent
+      CallInst *ci =
+          Builder.CreateCall(m_assert_if, {CS.getArgument(0), CS.getArgument(1)});
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/PromoteVerifierCalls.cc:185]: the arguments are unlikely to be `i1`. They would be a `zext` of `i1` to something expected by the function. See how `verifier.assume` is lifted.

File: lib/seahorn/BvOpSem2.cc [hunk 4]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
     setValue(*CS.getInstruction(), res);
   }
 
+  void visitAssertIf(CallSite CS) {
+    Expr ante = lookup(*CS.getArgument(0));
+    Expr conseq = lookup(*CS.getArgument(1));
+    m_ctx.m_z3_solver->reset();
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:561]: as a rule, we don't access fields of objects directly. Use a getter method.

File: lib/seahorn/BvOpSem2.cc [hunk 5]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    m_ctx.m_z3_solver->assertExpr(ante);
+    boost::tribool res = m_ctx.m_z3_solver->solveAssuming(m_ctx.side());
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:563]: this seems wrong. `side` is a side-condition, not a literal.
I would expect
```
m_ctx.m_z3_solver->assertExpr(m_ctx.side());
m_ctx.m_z3_solver->solve();
```

File: lib/seahorn/BvOpSem2.cc [hunk 6]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    if (res == boost::tribool::true_value) {
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:564]: the goal of tribool is to hide that it is tribool. Don't use `true_value` directly. Instead:
```
if (res) {  /* true */ }
else if (!res) { /* false */ }
else { /* unknown */ }
```

File: lib/seahorn/BvOpSem2.cc [hunk 7]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+      LOG("opsem", INFO << "Antecedent is sat: " << *CS.getInstruction() << "\n");
+    } else if (res == boost::tribool::false_value) {
+      LOG("opsem", ERR << "Antecedent is unsat: " << *CS.getInstruction() << "\n");
+    } else if (res == boost::tribool::indeterminate_value) {
+      LOG("opsem", WARN << "Antecedent is indeterminate: " << *CS.getInstruction() << "\n");
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:569]: `INFO`, `ERR`, and `WARN` add their own `\n`

File: lib/seahorn/BvOpSem2.cc [hunk 8]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    }
+    m_ctx.m_z3_solver->reset();
+    m_ctx.m_z3_solver->assertExpr(conseq);
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:572]: antecedent is missing!!!

File: lib/seahorn/BvOpSem2.cc [hunk 9]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    res = m_ctx.m_z3_solver->solveAssuming(m_ctx.side());
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:573]: see comment on `side()` before

File: lib/seahorn/BvOpSem2.cc [hunk 10]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    if (res == boost::tribool::true_value) {
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:574]: see comment on using `tribool` before

File: lib/seahorn/BvOpSem2.cc [hunk 11]
```diff
@@ -553,6 +555,32 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+      LOG("opsem", INFO << "Consequent is sat: " << *CS.getInstruction() << "\n");
+    } else if (res == boost::tribool::false_value) {
+      LOG("opsem", ERR << "Consequent is unsat: " << *CS.getInstruction() << "\n");
+    } else if (res == boost::tribool::indeterminate_value) {
+      LOG("opsem", WARN << "Consequent is indeterminate: " << *CS.getInstruction() << "\n");
+    }
+    setValue(*CS.getInstruction(), m_ctx.alu().getTrue());
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:581]: `assert.if` should be `void`. There is no return value.

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc (hunk 3 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/PromoteVerifierCalls.cc]: done

File: lib/Analysis/SeaBuiltinsInfo.cc (hunk 2 above)
Comment (priyasiddharth) [reply lib/Analysis/SeaBuiltinsInfo.cc]: ack.

File: include/seahorn/Analysis/SeaBuiltinsInfo.hh (hunk 1 above)
Comment (priyasiddharth) [reply include/seahorn/Analysis/SeaBuiltinsInfo.hh:20]: For now i am adding a verifier.assert.not and mapping to sea.assert.if... we can add it i run into problems with the extra negation instruction.

File: lib/seahorn/BvOpSem2.cc (hunk 4 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: Agreed. That is why this is a draft :). Fixed it now.

File: lib/seahorn/BvOpSem2.cc (hunk 5 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: Fixed

File: lib/seahorn/BvOpSem2.cc (hunk 6 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: Fixed

File: lib/seahorn/BvOpSem2.cc (hunk 8 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: Fixed.

File: lib/seahorn/BvOpSem2.cc (hunk 9 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: ack

File: lib/seahorn/BvOpSem2.cc (hunk 10 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: ack

File: lib/seahorn/BvOpSem2.cc (hunk 11 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: Fixed.

File: lib/seahorn/BvOpSem2.cc (hunk 7 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: I have noted to fix this in the project. Also added a task for separating out unsat/undet printing

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc [hunk 12]
```diff
@@ -173,6 +179,24 @@ bool PromoteVerifierCalls::runOnFunction(Function &F) {
       if (cg)
         (*cg)[&F]->addCalledFunction(ci, (*cg)[ci->getCalledFunction()]);
 
+      I.replaceAllUsesWith(ci);
+      toKill.push_back(&I);
+    } else if (fn && (fn->getName().equals(
+                         "__VERIFIER_assert_if"))) { // sea_assert_if is the
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/PromoteVerifierCalls.cc:185]: should not you also handle `sea_assert_if`

File: lib/seahorn/BvOpSem2.cc [hunk 13]
```diff
@@ -1804,7 +1845,8 @@ Bv2OpSemContext::Bv2OpSemContext(SymStore &values, ExprVector &side,
       m_fparams(o.m_fparams), m_ignored(o.m_ignored),
       m_registers(o.m_registers), m_alu(nullptr), m_memManager(nullptr),
       m_parent(&o), zeroE(o.zeroE), oneE(o.oneE), m_z3(o.m_z3),
-      m_z3_simplifier(o.m_z3_simplifier) {
+      m_z3_simplifier(o.m_z3_simplifier),
+      m_z3_solver(std::move(o.m_z3_solver)) {
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:1849]: probably just want to re-create the solver. Definitely not move it

File: test/opsem/lit.cfg [hunk 14]
```diff
@@ -61,6 +68,10 @@ seabmc = [seahorn_cmd, '--horn-bmc-engine=mono',
 seasmt = [seahorn_cmd, '--horn-bmc-engine=mono',   
           '--horn-bmc', '--horn-bv2=true', '--log=opsem',
           '--keep-shadows=true' ]
+sea = [sea_cmd, 'bpf', '--horn-bmc-engine=mono',
```
Comment (agurfinkel) [inline test/opsem/lit.cfg:71]: we should start using yama command in tests and store configuration file inside test directory

**Comment by agurfinkel** [comment]
looking good. does it perform well on our bigger bench set?

File: lib/seahorn/BvOpSem2.cc [hunk 15]
```diff
@@ -555,6 +560,42 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
 
+  // TODO: visit.assert.if.not is not supported
+  void visitAssertIf(CallSite CS) {
+    auto I = CS.getInstruction();
+    Expr ante = lookup(*CS.getArgument(0));
+    Expr conseq = lookup(*CS.getArgument(1));
+    tribool anteRes = solveWithConstraints(ante);
+    // if ante is unsat then report false and bail out
+    if (!anteRes) {
+      LOG("opsem", ERR << "Antecedent is unsat/undet: " << *I << "\n");
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:571]: `unsat/unknown` -- user facing terminology is `sat`, `unsat`, `unknown`

File: lib/seahorn/BvOpSem2.cc [hunk 16]
```diff
@@ -555,6 +560,42 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
     setValue(*CS.getInstruction(), res);
   }
 
+  // TODO: visit.assert.if.not is not supported
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:563]: move comment inside the function. Typo `visit.assert...` should be just `assert...`

File: lib/seahorn/BvOpSem2.cc [hunk 17]
```diff
@@ -555,6 +560,42 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    }
+    Expr e = boolop::land(ante, boolop::lneg(conseq));
+    tribool conseqRes = solveWithConstraints(e);
+    if (conseqRes) {
+      LOG("opsem", ERR << "Consequent is sat/undet: " << *I << "\n");
+    }
+  }
+
+  tribool solveWithConstraints(const Expr &solveFor) const {
+    m_ctx.solver().reset();
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:582]: nothing to do now, but this function will need to be improved once we start measuring for performance

File: lib/seahorn/BvOpSem2.cc [hunk 18]
```diff
@@ -555,6 +560,42 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    tribool anteRes = solveWithConstraints(ante);
+    // if ante is unsat then report false and bail out
+    if (!anteRes) {
+      LOG("opsem", ERR << "Antecedent is unsat/undet: " << *I << "\n");
+      return; // return early
+    }
+    Expr e = boolop::land(ante, boolop::lneg(conseq));
+    tribool conseqRes = solveWithConstraints(e);
+    if (conseqRes) {
+      LOG("opsem", ERR << "Consequent is sat/undet: " << *I << "\n");
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:577]: need options for vacuity and for incremental checking. Will need to time how long they take. We might need vacuity + final check, but not vacuity + incremental + final

File: test/opsem/lit.cfg [hunk 19]
```diff
@@ -61,6 +68,10 @@ seabmc = [seahorn_cmd, '--horn-bmc-engine=mono',
...
+       '--horn-bmc', '--horn-bv2=true', '--log=opsem',
+       '--keep-shadows=true']
```
Comment (agurfinkel) [inline test/opsem/lit.cfg:73]: keep shadows should never be passed to `sea` during normal operation

File: lib/seahorn/BvOpSem2.cc (hunk 17 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:582]: added optimize comment

File: lib/seahorn/BvOpSem2.cc (hunk 16 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:563]: ack.

File: include/seahorn/seahorn.h [hunk 20]
```diff
@@ -19,18 +19,31 @@ extern void __VERIFIER_error(void);
...
+extern void __VERIFIER_assert_not(bool);
+extern void __VERIFIER_assert_if(bool, bool);
 /**
    Returns TRUE if \p offset number of bytes of \p ptr are allocated
 
     Requires support from the memory manager. Might be interpreted to always
     return TRUE or FALSE if the memory manager does not support it.
  */
 extern bool sea_is_dereferenceable(void *ptr, intptr_t offset);
+extern bool sea_assert_if(bool, bool);
```
Comment (agurfinkel) [inline include/seahorn/seahorn.h:32]: is `sea_assert_if` used? Remove if not used, make `void` if used

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc [hunk 21]
```diff
@@ -99,6 +103,8 @@ bool PromoteVerifierCalls::runOnFunction(Function &F) {
       fn = dyn_cast<const Function>(CS.getCalledValue()->stripPointerCasts());
 
     if (fn && (fn->getName().equals("__VERIFIER_assume") ||
+               fn->getName().equals("__VERIFIER_assert") ||
+               fn->getName().equals("__VERIFIER_assert_not") ||
                fn->getName().equals("DISABLED__VERIFIER_assert") ||
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/PromoteVerifierCalls.cc:108]: remove `DISABLED` line. Looks like a left over from something else

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 22]
```diff
@@ -48,48 +56,68 @@ class UnifyAssumesPass : public ModulePass {
...
+  case SeaBuiltinsOp::ASSERT:
+  case SeaBuiltinsOp::ASSERT_NOT:
+    return true;
+  }
+}
+
 /// Find a function exit basic block.  Assumes that the function has
 /// a unique block with return instruction
-llvm::BasicBlock *findExitBlock(llvm::Function &F) {
+llvm::BasicBlock *UnifyAssumesPass::findExitBlock(llvm::Function &F) {
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc:82]: why is this function a method? used to be just a local function.

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 23]
```diff
@@ -48,48 +56,68 @@ class UnifyAssumesPass : public ModulePass {
...
   }
   return nullptr;
 }
 
 bool UnifyAssumesPass::runOnFunction(Function &F) {
   Module &M = *F.getParent();
-  auto &SBI = getAnalysis<seahorn::SeaBuiltinsInfoWrapperPass>().getSBI();
-  auto *assumeFn = SBI.mkSeaBuiltinFn(seahorn::SeaBuiltinsOp::ASSUME, M);
+  m_SBI = &getAnalysis<seahorn::SeaBuiltinsInfoWrapperPass>().getSBI();
+  assert(m_SBI);
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc:102]: no need to assert. this assert is always true.

File: lib/Transforms/Scalar/UnifyAssumes.cc [hunk 24]
```diff
@@ -48,48 +56,68 @@ class UnifyAssumesPass : public ModulePass {
...
+      if (isAssumeCall(*ci)) {
         assumes.push_back(ci);
+      } else if (isAssertCall(*ci)) {
+        asserts.push_back(ci);
+      }
     }
   }
-  // -- nothing to unify if there is only one assume already
-  if (assumes.size() <= 1)
+  if (assumes.size() < 1) {
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/UnifyAssumes.cc:116]: remove. We agreed it is not a warning. We will deal with `verifier.assert` in opsem

**Comment by agurfinkel** [comment]
Very minor things. One more: commit history contains merge pull request. Don't need that in the history.

ready to merge after the little things are addressed

**Comment by agurfinkel** [comment]
oh, and travis build is failing :)

File: test/opsem/lit.cfg (hunk 19 above)
Comment (priyasiddharth) [reply test/opsem/lit.cfg:73]: ack

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 22 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc:82]: In the previous review you asked me to change SBI to a class member and use it in one of the methods. I assumed that you wanted to pass this to all functions that use SBI and making everything a method was the natural conclusion i drew. Also the mix of local functions and methods seemed arbitrary AFAICT. Is there a reason why this can't be a function?

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 24 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc:116]: done

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 23 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc:102]: done

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc (hunk 21 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/PromoteVerifierCalls.cc:108]: done

File: include/seahorn/seahorn.h (hunk 20 above)
Comment (priyasiddharth) [reply include/seahorn/seahorn.h:32]: used for "unit testing" sea.assert.if in tests/opsem. Fixed.

**Comment by priyasiddharth** [comment]
I believe I have addressed all comments. Will now focus on CI failure.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 22 above)
Comment (agurfinkel) [reply lib/Transforms/Scalar/UnifyAssumes.cc:82]: `findExitBlock` does not use `SBI`. If it does, it should be a method.

Better use the simplest concept that works. If a function is just a function, it should not be part of a class. This is not Java. Often class will have functions that only the class uses, these can be hidden inside cpp files. Sometimes a function, like `findExitBlock` is waiting to be promoted to be used by other clients as well.

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 24 above)
Comment (agurfinkel) [reply lib/Transforms/Scalar/UnifyAssumes.cc:115]: `assumes.size() < 1` is the same as `assumes.size() == 0` same as `assumes.empty()`

File: lib/seahorn/BvOpSem2.cc (hunk 18 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: done

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 24 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc:114]: fixed

File: lib/Transforms/Scalar/UnifyAssumes.cc (hunk 22 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/UnifyAssumes.cc]: reverted change to `findExitBlock`

**Comment by priyasiddharth** [comment]
The new tests are not part of the CI. The PR is ready to be merged from my end.

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc [hunk 25]
```diff
@@ -159,7 +168,7 @@ bool PromoteVerifierCalls::runOnFunction(Function &F) {
 
       IRBuilder<> Builder(F.getContext());
       Builder.SetInsertPoint(&I);
-      CallInst *ci = Builder.CreateCall(m_failureFn);
+      CallInst *ci = Builder.CreateCall(m_errorFn);
```
Comment (agurfinkel) [inline lib/Transforms/Scalar/PromoteVerifierCalls.cc:171]: why was this changed from `m_failureFn` to `m_errorFn`?

File: lib/seahorn/BvOpSem2.cc [hunk 26]
```diff
@@ -555,6 +579,53 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+
+    boost::tribool res = m_ctx.solver().solve();
+    return res;
+  }
+
+  void visitAssertStmt(CallSite CS) {
+    // This is called when UnifyAssumes does not replace
+    // verifier.assert with sea.assert.if e.g.
+    // when no assumes is found, as a workaround do the following
+    // TODO: add logic to delegate to sea.assert.if(true, conseq)
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:624]: hope this todo is marked in the project as an issue and will not be hidden in here.

File: lib/Transforms/Scalar/PromoteVerifierCalls.cc (hunk 25 above)
Comment (priyasiddharth) [reply lib/Transforms/Scalar/PromoteVerifierCalls.cc:171]: Sorry i was trying auto format on save in emacs... it messed up things... and i missed reverting this.

File: lib/seahorn/BvOpSem2.cc (hunk 26 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:624]: done

**Comment by priyasiddharth** [comment]
Responded to latest comments.
